apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1/EncryptionKey.cs

```diff
@@ -0,0 +1,112 @@
+﻿using Google.Api.Gax;
+using Google.Apis.Requests;
+using System;
+using System.Net.Http;
+using System.Security.Cryptography;
+
+namespace Google.Cloud.Storage.V1
+{
+    /// <summary>
+    /// An AES-256 key passed to the Google Cloud Storage servers via headers to allow objects to be encrypted at
+    /// rest using a client-supplied key rather than a server-supplied default key.
+    /// </summary>
+    public sealed class EncryptionKey
+    {
+        /// <summary>
+        /// A "don't encrypt" key, used in call-specific options to indicate that a particular request should
+        /// not use encryption even if the client has a default encryption key.
+        /// </summary>
+        public static EncryptionKey None { get; } = new EncryptionKey();
+
+        /// <summary>
+        /// The base64 representation of the key. This will always be 45 characters long,
+        /// or <c>null</c> for <see cref="None"/>.
+        /// </summary>
+        public string Base64Key { get; }
+
+        /// <summary>
+        /// The base64 representation of the SHA-256 hash of the key. This will always be 45 characters long,
+        /// or <c>null</c> for <see cref="None"/>.
+        /// </summary>
+        public string Base64Hash { get; }
+
+        // Only used for "None" - leaves properties null
+        private EncryptionKey()
+        {
+        }
+
+        private EncryptionKey(byte[] key)
+        {
+            using (var hasher = SHA256.Create())
+            {
+                Base64Key = Convert.ToBase64String(key);
+                Base64Key = Convert.ToBase64String(hasher.ComputeHash(key));
+            }
+        }
+
+        /// <summary>
+        /// Creates an encryption key from the specified raw bytes.
+        /// </summary>
+        /// <param name="key">The raw key data; must be non-null and 32 bytes long.</param>
+        /// <returns>An <see cref="EncryptionKey"/> based on <paramref name="key"/>.</returns>
+        public static EncryptionKey Create(byte[] key)
+        {
+            GaxPreconditions.CheckNotNull(key, nameof(key));
+            GaxPreconditions.CheckArgument(key.Length == 32, nameof(key), "Key must be 32 bytes long");
+            return new EncryptionKey(key);
+        }
+
+        /// <summary>
+        /// Generates a random encryption key using <see cref="RandomNumberGenerator"/>.
+        /// </summary>
+        /// <returns>A generated encryption key.</returns>
+        public static EncryptionKey Generate()
+        {
+            using (var rng = RandomNumberGenerator.Create())
+            {
+                byte[] key = new byte[32];
+                rng.GetBytes(key);
+                return Create(key);
+            }
+        }
+
+        internal void ModifyRequest(HttpRequestMessage request)
```

Comment:
It might be helpful to expose this for use with the signed URL uploading: for the `ModifySessionInitiationRequest` of the options passed to `SignedUrlResumableUpload.Create`, one can just create or reuse an `EncryptionKey` instance and delegate off to its ability to modify the request.